src/test/java/study/CalculatorTest.java

```diff
@@ -0,0 +1,212 @@
+    })
+    void multiplyTest(String input, long expectResult) {
+        assertThat(calculate(input)).isEqualTo(expectResult);
+    }
+
+    @ParameterizedTest
+    @DisplayName("'/' 연산 테스트")
+    @CsvSource({"0 / 1,0", "1 / 1,1", "2 / 1,2", "144 / 12,12", // 여러 자릿수
+            "0 / 1,0", "10 / 2 / 5,1", "300 / 3 / 10,10" // 여러 수
+    })
+    void divisionTest(String input, long expectResult) {
+        assertThat(calculate(input)).isEqualTo(expectResult);
+    }
+
+    @Test
+    @DisplayName("'/' 연산 에서 정수로 떨어지지 않는 경우 예외처리")
+    void divideResultNotIntegerException() {
+        Assertions.assertThatExceptionOfType(IllegalArgumentException.class) //
+                .isThrownBy(() -> calculate("3 / 2")) //
+                .withMessage("나눗셈은 정수로 떨어져야 합니다.");
+    }
+
+    @Test
+    @DisplayName("사칙연산 연산자가 아닌경우 예외처리")
+    void invalidOperator() {
+        Assertions.assertThatExceptionOfType(IllegalArgumentException.class) //
+                .isThrownBy(() -> calculate("3 ~ 2")) //
+                .withMessage("올바른 연산자가 아닙니다.");
+    }
+
+    @Test
+    @DisplayName("여러 연산이 같이 있는 경우 앞쪽 연산부터 수행한다")
+    void complexTest() {
+        assertThat(calculate("2 + 3 * 4 / 2")).isEqualTo(10);
+    }
+
+    private long calculate(String input) {
+        if (Objects.isNull(input) || input.isEmpty()) {
+            throw new IllegalArgumentException();
+        }
+
+        List<Parsed> parsingResult = new ArrayList<>();
+
+        do {
+            Parsed parsed = parse(input);
+            parsingResult.add(parsed);
+            input = parsed.remain;
+        } while (!Objects.isNull(input));
+
+        long result = 0;
+        for (Parsed parsed : parsingResult) {
+            result = parsed.calculateWith(result);
+        }
+        return result;
+    }
+
+    private Parsed parse(String input) {
+        try {
+            String lhs = extractPrefixNumbers(input);
+            String operator = input.substring(lhs.length() + SPACE.length(), lhs.length() + SPACE.length() + OPERATOR_LENGTH);
+            String rhs = input.substring(input.indexOf(operator) + OPERATOR_LENGTH + SPACE.length());
+            String remain = null;
+
+            String tmp = rhs;
+            if (isNotNumber(rhs)) {
+                rhs = extractPrefixNumbers(tmp);
+                remain = tmp.substring(rhs.length());
+            }
+
+
+            return new Parsed(lhs, operator, rhs, remain);
+        } catch (StringIndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("연산자 사이에는 빈 공간이 한칸 있어야 합니다.", e);
+        }
+    }
+
+    private String extractPrefixNumbers(String input) {
+        String result = input.substring(0, input.indexOf(SPACE));
+        if (isNotNumber(result) && !result.isEmpty()) {
+            throw new IllegalArgumentException("연산자 사이에는 빈 공간이 한칸 있어야 합니다.");
+        }
+        return result;
+    }
+
+    private boolean isNotNumber(String input) {
+        return !Pattern.matches("\\d+", input);
```

Comment:
현재 메서드 호출할때마다 Pattern.matches를 호출하고 있네요 :)


Pattern.matches 내부 구현을 보면
```
// Pattern
public static boolean matches(String regex, CharSequence input) {
        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(input);
        return m.matches();
}
```
해당 메서드가 호출될때마다 패턴을 컴파일하도록 되어있습니다. Pattern.compile()은 성능이 비싸기 때문에
매번 메서드 호출시 성능이 비싼 Pattern 인스턴스를 생성하고 Pattern 인스턴스는 한번 쓰고 버려지기 때문에 성능적으로 좋지 않습니다.

자바 doc에서도 한번 컴파일 후 재사용 할것을 권장하고 있습니다.
```
If a pattern is to be used multiple times, compiling it once and reusing
it will be more efficient than invoking this method each time
```

불필요한 객체 생성을 피하도록 개선하면 좋을 것 같아요 :)

Reply:
저는 상황에 따라 다르다고 생각합니다. 복잡한 정규식이라면 컴파일 해 놓는게 당연하지만 단순한 경우에는 큰 의미 없다고 생각해요. 특히 요즘처럼 좋은 서버환경에서는 더욱 성능개선의 효과는 떨어지며 오히려 가독성을 떨어뜨리는 요인이 된다고 생각합니다. 
저는 변경하지 않는게 좋겠다고 생각해요 😀

Reply:
광수님 생각도 충분히 일리있는 말씀이라고 생각합니다 :)
말씀해주신대로 요즘 하드는 좋은 성능을 가지고 있기에 크게 문제는 될 일이 없다고 생각해요ㅎㅎ
다만 위에서 언급했던 내용중 자바 Doc에서 한번 컴파일 후 재사용 할 것을 권장하고 있사오니
해당 내용만 인지하고 넘어가주시면 좋을 것 같아요 :)